src/Sylius/Bundle/CoreBundle/Validator/Constraints/CatalogPromotionRulesValidityValidator.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Validator\Constraints;
+
+use Sylius\Component\Core\Repository\ProductVariantRepositoryInterface;
+use Sylius\Component\Promotion\Model\CatalogPromotionRuleInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+final class CatalogPromotionRulesValidityValidator extends ConstraintValidator
+{
+    private ProductVariantRepositoryInterface $variantRepository;
+
+    public function __construct(ProductVariantRepositoryInterface $variantRepository)
+    {
+        $this->variantRepository = $variantRepository;
+    }
+
+    public function validate($value, Constraint $constraint): void
+    {
+        /** @var CatalogPromotionRuleInterface $rule */
```

Comment:
If we assert a proper constraint type (`CatalogPromotionRulesValidity`) at the beginning of the function, it would be easier to use it later (and we would have an autocomplete from IDE 🚀)